[FEATURE] : [경매 종료가 임박한 진행중인 경매 목록 조회]

na0th opened the pull request:
1. 경매 종료가 임박한 진행중인 경매 목록 조회 추가

미흡 : 성능 개선 못함,
/auction으로 엔드포인트 합쳐야 하는데 못했음,
테스트쪽 리뷰 받아야 함.

---
name: Pull Request
about: '풀 리퀘스트를 제출합니다.'
title: "[FEATURE] - [기능명] 또는 [FIX] - [버그 설명]"
labels: ''
assignees: ''

---
* #9 에 대한 Issue 을 중점으로 구현
## 🆕 기능 추가 / 🔧 버그 수정

경매 종료가 임박했고, 진행 중인 경매 목록을 조회합니다.

## 📋 변경 사항


## 🔍 테스트 사항
Controller 테스트 미구현

Repository, Service 테스트 미흡, 꼼꼼한 리뷰 필수 

## 📄 관련 문서
- 변경 사항과 관련된 문서나 링크가 있다면 여기에 추가해주세요.

## 📝 추가 사항
/auction으로 엔드포인트 합쳐야 하는데 못했습니다.

na0th commented on src/main/java/com/tasksprints/auction/api/auction/AuctionController.java:
```diff
@@ -82,17 +82,23 @@ public ResponseEntity<ApiResult<String>> getAuctionStatus(@PathVariable Long auc
-    List<AuctionResponse> auctions = auctionService.getAuctionsByFilter(productCategory,auctionCategory);
-    return ResponseEntity.ok(ApiResult.success(ApiResponseMessages.ALL_AUCTIONS_RETRIEVED, auctions));
-}
-
```
줄맞춤 때문에 TAP 키 한 번 쳣습니다.

yachimiya replied:
자주 참여하지 못해서 죄송합니다 :cry: 

이런 부분은 리뷰어가 좀 더 핵심적인 부분에 집중할 수 있도록 checkstyle 같은 정적 분석 툴로 검사하거나 CI 파이프라인에서 걸러낼 수 있으면 좋을 것 같네요! :thinking:

na0th commented on src/main/java/com/tasksprints/auction/api/auction/AuctionController.java:
```diff
@@ -82,17 +82,23 @@ public ResponseEntity<ApiResult<String>> getAuctionStatus(@PathVariable Long auc
+    ) {
+        List<AuctionResponse> auctions = auctionService.getAuctionsByFilter(productCategory,auctionCategory);
+        return ResponseEntity.ok(ApiResult.success(ApiResponseMessages.ALL_AUCTIONS_RETRIEVED, auctions));
+    }
```
엔드포인트 /auction으로 합쳐야 했는데, 잘 모르겠어서 분리했습니다. (리팩토링 요청)

na0th commented on src/test/java/com/tasksprints/auction/domain/auction/service/AuctionServiceImplTest.java:
```diff
@@ -363,4 +414,14 @@ private Auction createAuction(Long auctionId, User seller, AuctionStatus status)
+                    status,
+                    seller
+            );
+        }
```
endTime을 직접 설정해야 해서, 기존 생성자 코드를 그대로 두고, 아래에 파라미터만 1개 늘어난 생성자를 추가했습니다.

polyglot-k reviewed (changes requested):
일단 전체적으로 작업 진행해주셔서 감사합니다! 일단 현재 Service 내부 로직은 의도에 맞게 잘 설계된 것 같습니다.

아쉬운 부분은 CriteriaRepository 는 전체 조회에서의 모든 filter 를 포함할 수 있도록 하려고 구성해 놓은 것이라서, 이를 좀 더 활용했으면 좋았을 것 같습니다.

인자를 추가하는 형태로 진행했어도 괜찮았을 것 같습니다.

Controller 를 새롭게 추가하는 것은, 지금 고민해봤을 때, 마감임박 요소만을 빠르게 불러오는 것을 만들어야하는지에 대한 고민을 가지고 있습니다. 일단 만들게 되었을 때는, 해당 기능을 통해서 `fit`하게 가지고 올 수 있다는 장점이 있는데, 단점은 restful 하지 못할 수도 있다라는 점이 있긴 합니다. 그래서 해당 부분에 대한 유지보수가 조금 필요해 보입니다!

yachimiya commented on src/main/java/com/tasksprints/auction/domain/auction/service/AuctionServiceImpl.java:
```diff
@@ -135,4 +136,32 @@ public List<AuctionResponse> getAuctionsByFilter(AuctionRequest.ProductCategoryP
+    @Override
+    public List<AuctionResponse> getAuctionsByEndTimeBetweenOrderByEndTimeAsc() {
+        // auction의 endTime까지 24시간 이하로 남은 진행중인 경매 목록 조회
+        LocalDateTime now = LocalDateTime.now();
```
`LocalDateTime.now()`가 환경에 따라서 깨지기 쉬운 메서드인 것 같아서(예: 배포 환경의 TZ에 따라서 예상하지 못한 결과를 얻을 수 있는 등) 이 부분은 주의가 필요할 것 같네요! :thinking: 

- https://code-run.tistory.com/m/82

KNU-K replied:
해당 부분은 문서화하여 공유하면 좋을 것 같습니다

yachimiya commented on src/test/java/com/tasksprints/auction/domain/auction/repository/AuctionRepositoryTest.java:
```diff
@@ -122,6 +122,34 @@ public void testFindAllUsingFilter() {
+    @DisplayName("경매 마감 시간까지 24시간 이하로 남은 경매 목록 조회")
+    public void testFindAuctionsByEndTimeBetweenOrderByEndTimeAsc() {
+        //given
+        Auction auction1 = createAuction(seller, LocalDateTime.now().plusHours(21), AuctionCategory.PUBLIC_PAID, AuctionStatus.ACTIVE);
+        Auction auction2 = createAuction(seller, LocalDateTime.now().plusHours(22), AuctionCategory.PUBLIC_PAID, AuctionStatus.ACTIVE);
+        Auction auction3 = createAuction(seller, LocalDateTime.now().plusHours(23), AuctionCategory.PUBLIC_PAID, AuctionStatus.ACTIVE);
+        Auction auction4 = createAuction(seller, LocalDateTime.now().plusHours(48), AuctionCategory.PUBLIC_PAID, AuctionStatus.ACTIVE);
+        Auction auction5 = createAuction(seller, LocalDateTime.now().plusHours(20), AuctionCategory.PUBLIC_PAID, AuctionStatus.PENDING);
```
이런 부분은 테스트에 사용되는 데이터를 빠르게 셋업할 수 있도록 별도의 테스트 픽스처가 있으면 좋을 것 같네요. 오브젝트 마더나 테스트 빌더 등을 통해서 테스트에 사용되는 합리적인 기본값을 설정해두고, 테스트의 대상이 되는 데이터만 외부로 노출시키는 것도 괜찮을 것 같습니다 :thinking:

yachimiya commented on src/test/java/com/tasksprints/auction/domain/auction/repository/AuctionRepositoryTest.java:
```diff
@@ -122,6 +122,34 @@ public void testFindAllUsingFilter() {
+        // when
+        List<Auction> result = auctionRepository.getAuctionsEndWith24Hours(now, next24Hours, AuctionStatus.ACTIVE);
+        //then
+        assertThat(result).hasSize(3);
+        // 오름차순 정렬 되었는지 확인
+        LocalDateTime previousEndTime = null;
+            for (Auction auction : result) {
+                if (previousEndTime != null) {
+                    assertThat(auction.getEndTime()).isAfterOrEqualTo(previousEndTime);
+                }
+                previousEndTime = auction.getEndTime();
+            }
```
이런 부분은 테스트 실패 시 빠르고 정확한 피드백을 받을 수 있도록 `assertAll`이나 `assertSoftly` 같이 soft assertion을 이용하면 괜찮겠다 생각이 들었습니다 :smile:

yachimiya commented on src/test/java/com/tasksprints/auction/domain/auction/repository/AuctionRepositoryTest.java:
```diff
@@ -122,6 +122,34 @@ public void testFindAllUsingFilter() {
+
+        auctionRepository.saveAll(List.of(auction1, auction2, auction3, auction4, auction5));
+
+        LocalDateTime now = LocalDateTime.now();
+        LocalDateTime next24Hours = LocalDateTime.now().plusHours(24);
```
이 부분 자체는 문제가 없을 수 있지만 테스트 데이터가 동적인 것보다는 고정적이면 외부 환경이나 미묘한 타이밍으로 인한 문제를 약간이나마 배제할 수 있다는 점이 있을 것 같네요 :thinking: 

- https://jojoldu.tistory.com/416

na0th replied:
꼼꼼한 리뷰 감사합니다!! 많이 미숙하지만 지적해주신 부분 검색해서 보완하겠습니다~

thornscript reviewed (commented):
자주 참여하지 못해서 죄송합니다 :cry: 

수고하셨습니다 :+1:

na0th committed:
Feat : 경매 종료가 임박한 진행중인 경매 목록 조회

1. 경매 종료가 임박한 진행중인 경매 목록 조회 추가

미흡 : 성능 개선 못함,
/auction으로 엔드포인트 합쳐야 하는데 못했음,
테스트쪽 리뷰 받아야 함.

na0th committed:
Merge branch 'feat-pr' of https://github.com/na0th/auction into feat-pr

na0th committed:
style : 코드 포맷터 빌드

코드 포맷터 빌드

na0th committed:
Refactor : 마감이 얼마 안남은 경매 조회 메서드에 파라미터 추가

1. 마감이 얼마 안 남은 경매 목록 조회 메서드의 파라미터가 없던 것에서 now, next24Hours 추가

na0th committed:
Test: 마감 기한 얼마 안 남은 경매 목록 조회 : 테스트 코드 수정

1. 테스트 케이스의 LocalDateTime을 고정
2. 테스트 대상이 아닌 파라미터는 테스트 픽스처에 넣음
3. assertAll로 컬렉션인 경우 끝까지 assertion하고 결과를 보여주도록 함

KNU-K commented on src/test/java/com/tasksprints/auction/domain/auction/repository/AuctionRepositoryTest.java:
```diff
@@ -125,14 +148,38 @@ public void testFindAllUsingFilter() {
+        LocalDateTime fixedNow = LocalDateTime.of(2024, 9, 1, 10, 0);
+        LocalDateTime next24Hours = fixedNow.plusHours(24);
+
+        Auction auction1 = createAuction(fixedNow.plusHours(21), AuctionStatus.ACTIVE);
+        Auction auction2 = createAuction(fixedNow.plusHours(22), AuctionStatus.ACTIVE);
+        Auction auction3 = createAuction(fixedNow.plusHours(23), AuctionStatus.ACTIVE);
+        Auction auction4 = createAuction(fixedNow.plusHours(48), AuctionStatus.ACTIVE);
+        Auction auction5 = createAuction(fixedNow.plusHours(20), AuctionStatus.PENDING);
```
애초에 리스트로 만들면 더 직관적일 것 같아요

KNU-K commented on src/main/java/com/tasksprints/auction/domain/auction/repository/support/AuctionCriteriaRepositoryImpl.java:
```diff
@@ -29,4 +32,19 @@ public List<Auction> getAuctionsByFilters(ProductCategory productCategory,
+        return queryFactory
+                .selectFrom(auction)
+                .where(endTimeBetween.and(statusEquals))
+                .orderBy(auction.endTime.asc())
```
`order by`가 성능 저하의 병목지점이 될 수 있음에 유의 해야합니다! 이에 대한 방안을 고려해보고, 최선책으로 `order by`를 사용해야하는 근거를 가지고 오시면 좋을 것 같습니다.!

KNU-K replied:
+ 제가 이전에 만든 filter 함수에 합치는게 더 낫지않을까요? 충분히 합칠 수 있다고 생각이 됩니다.

na0th replied:
넵 노력하겠습니다

KNU-K commented on src/test/java/com/tasksprints/auction/domain/auction/repository/AuctionRepositoryTest.java:
```diff
@@ -46,6 +47,28 @@ public void setUp() {
     }
 
+    private Auction createAuction(User seller, AuctionCategory category, AuctionStatus status) {
+        return Auction.create(
+            LocalDateTime.now(),
+            LocalDateTime.now().plusDays(7),
+            BigDecimal.valueOf(100.00),
+            category,
+            status,
+            seller
+        );
+    }
+
+    private Auction createAuction(LocalDateTime endTime, AuctionStatus status) {
+        return Auction.create(
+            LocalDateTime.now(),
+            endTime,
+            BigDecimal.valueOf(100.00),
+            AuctionCategory.PUBLIC_PAID,
+            status,
+            seller
+        );
```
LocalDateTime 피드백에 따라 고정으로 변경해주세요

na0th replied:
넵,

KNU-K commented on src/test/java/com/tasksprints/auction/domain/auction/service/AuctionServiceImplTest.java:
```diff
@@ -69,6 +70,18 @@ private Auction createAuction(Long auctionId, User seller, AuctionStatus status)
+            .id(auctionId)
+            .auctionCategory(AuctionCategory.PUBLIC_PAID)
+            .auctionStatus(status)
+            .startTime(LocalDateTime.now())
```
여기도 고정으로 변경해주세요

na0th replied:
넵, 고정으로 변경하겠습니다

na0th committed:
test : 테스트 코드 수정

1. createAuction 중복 코드 리스트로 수정
2. LocalDateTime.now()를 고정 시간으로 수정
3. AuctionServiceImplTest 중복 코드 삭제

na0th committed:
refactor : endingSoon -> /auction 검색 필터로 엔드포인트 합치기

1: /auction에서 검색 조건별로 경매 목록을 조회한다.

미흡 : 파라미터를 SearchCondition이라는 DTO로 받아서 기존 테스트 다 깨져서 수정해야 함,
변경점에 대한 테스트 코드 없음.
